File: submissions/0695-max-area-of-island/solution.ts
function maxAreaOfIsland(grid: number[][]): number {
  let m = grid.length;
  let n = grid[0].length;
  let seen = [];
  let max = 0;  
    
  for(let i = 0; i < m; i++){
      seen.push(new Array(n).fill(false));
  }  
    
  let valid = (row, col) => {
      return row >= 0 && col >= 0 && row < m && col < n && grid[row][col] == 1;
  }
   
  let directions = [[0,1],[1,0],[0,-1],[-1,0]];
    
  let dfs = (row, col) => {
      let temp = 1; 
      for(const [dx, dy] of directions){
          let nextRow = row + dy;
          let nextCol = col + dx;
          if(valid(nextRow, nextCol) && !seen[nextRow][nextCol]){
              seen[nextRow][nextCol] = true;
              temp += dfs(nextRow, nextCol);
          }   
    } 
          
   return temp; 
  }  
      
  
  for(let row = 0; row < m; row++){
      for(let col = 0; col < n; col++){
          if(!seen[row][col] && grid[row][col] == 1){
              seen[row][col] = true;
              max = Math.max(max, dfs(row, col));
          }
      }
  }
  
  return max;  
  
};

File: submissions/0695-max-area-of-island/solution.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        seen = set()
        directions = {(0, 1), (0, -1), (1, 0), (-1, 0)}
        
        def dfs(row, col):
            if row < 0 or row >= m or col < 0 or col >= n or (row, col) in seen or grid[row][col] == 0:
                return 0
             
            seen.add((row, col))    
            area = 1  
            for dx, dy in directions:
                area += dfs(row + dx, col + dy)
            
            return area
        
        max_area = 0
        for i in range(m):
            for j in range(n):
                if (i, j) not in seen and grid[i][j] == 1:
                    max_area = max(max_area, dfs(i, j))
        
        return max_area
